docs/CBOR.md

```diff
@@ -0,0 +1,139 @@
+`SizePolicy` that can be passed into `CborOutput`, e.g.
+
+```scala
+val cborList = CborOutput.write(List(1, 2, 3), sizePolicy = SizePolicy.Required)
+```
+
+`SizePolicy` has three values:
+
+* `Required` - Collections are always written with explicit size in their CBOR representation. This may be slow for
+  collections that don't have a fast `.size` method, e.g. `List`
+* `Optional` - Only collections that have a fast `.size` method are written with explicit size, e.g. `Vector`. This is
+  the default behavior.
+* `Ignored` - All collections are written without explicit size
+
+## Using CBOR keys of arbitrary type
+
+`GenCodec` has a strong assumption that all object/map keys are `String`s, like in JSON. One of the consequences is that
+serializing a `Map[K, V]` with `GenCodec` requires an instance of `GenKeyCodec` for `K` so that keys can be converted to
+strings.
+
+However, CBOR has no such limitations. Map keys can be arbitrary CBOR values. There are some additional tools
+in `GenCodec` and `CborInput/Output` that can leverage this.
+
+### `CborKeyCodec`
+
+`CborInput.read` and `CborOutput.write` accept a `CborKeyCodec` parameter that defines translation between string-typed
+keys and arbitrary CBOR-encoded keys. You can use this e.g. to give a numeric label to every textual field, reducing the
+size of resulting CBOR.
+
+```scala
+val keyCodec = new CborKeyCodec {
+  def writeFieldKey(fieldName: String, output: CborOutput): Unit = fieldName match {
+    case "intField" => output.writeInt(1)
+    case "strField" => output.writeInt(2)
+    case name => output.writeString(name)
+  }
+
+  def readFieldKey(input: CborInput): String = input.readInitialByte().majorType match {
+    case MajorType.Unsigned => input.readInt() match {
+      case 1 => "intField"
+      case 2 => "strField"
+      case n => throw new ReadFailure(s"unknown CBOR field label: $n")
+    }
+    case _ => input.readString()
+  }
+}
+
+case class Thing(intField: Int, strField: String)
+object Thing extends HasGenCodec[Thing]
+
+val cborBytes: Array[Byte] = CborOutput.write(Thing(42, "foo"), keyCodec)
+```
+
+This solution is nice when field names are often reused across multiple case classes that end up being serialized into
+final CBOR. Other than that, this is somewhat clunky and does not give you precise control on a level of each case
+class.
+
+### `HasCborCodec` and raw CBOR keys
+
+In place of `HasGenCodec`, you can use `HasCborCodec`. It generates an instance of `GenCodec` that is optimized for
+CBOR, which means two things:
+
+* if your case class contains a field typed as `Map[K, V]`, map keys will be serialized directly into CBOR rather than
+  converted into strings with `GenKeyCodec` (which would happen for standard `GenCodec`).
+* you can use `@cborKey` and `@cborDiscriminator` annotations in your case class & sealed hierarchy definitions
+
+Example:
+
+```scala
+@cborDiscriminator(0) sealed trait UnionData
```

Comment:
Let me know if I understand this correctly:
The `@cborDiscriminator` and `@cborKey` aren't tied to each other, you don't have  to use `@cborKey` when using a `@cborDiscriminator`.

I was confused for a few seconds when I saw `@cborKey` used in every variant of the sealed class. I think the documentation would be clearer id these two concepts were presented separatedly, or one of the variant class didn't use the `@cborKey` annotation.

Also, what is the difference between using a `String` `cborKey` and the standard GenCodec annotation for changing name of a serialized field (can't remember it's name)?

Reply:
> The @cborDiscriminator and @cborKey aren't tied to each other, you don't have to use @cborKey when using a @cborDiscriminator.
Yes.

> I was confused for a few seconds when I saw @cborKey used in every variant of the sealed class. I think the documentation would be clearer id these two concepts were presented separatedly, or one of the variant class didn't use the @cborKey annotation.

Elaborate pls how you imagine it, I'm not sure what you mean (what do you mean by "variants", in particular?)

`@cborDiscriminator` is effectively a `@cborKey` for the generated discriminator field. Because the field is generated, there's no case class field to apply `@cborKey` to. Therefore a separate annotation is needed.

> Also, what is the difference between using a String cborKey and the standard GenCodec annotation for changing name of a serialized field (can't remember it's name)?

Effectively no difference except that `@cborKey` will only be honored by `CborInput`/`CborOutput` while `@name` works with every `Input`/`Output` implementation.

Reply:
by `variants` i mean the `case class`es that inherit from the `sealed trait`. What is the correct name?


> Elaborate pls how you imagine it
Something like this, so it doesn't look like there is a need to use "discriminator" and "field rename" `cborKeys` simultaneously
```
object UnionData extends HasCborCodec[UnionData] {
  @cborKey(1) case class Textual(@cborKey(1) txt: String) extends UnionData
  @cborKey(2) case class Numeric(@cborKey(true) num: Int) extends UnionData
  @cborKey(3) case class Mapping(map: Map[Int, String]) extends UnionData
}
```

Anyway, ofiginal is LGTM if you think this is inferior

Reply:
Ok, so the problem is that I shoved all the features in a single example at once?

Reply:
Yes, at least it confused me (YMMV)

Reply:
TBH it makes sense to set `@cborDiscriminator` if you're also setting `@cborKey`s so I wouldn't mind if people did things like in this example.